csss-site/src/tools/lib/generate_document.py

```diff
@@ -0,0 +1,222 @@
+                annotation_obj = annotation.get_object()
+                annotation_type = annotation_obj['/T']
+                if annotation_type == "Check Box1":
+                    found_check_box_1 = True
+                elif annotation_type == "Check Box3":
+                    found_check_box_3 = True
+
+                if annotation_type == "Check Box1" and data.receive_kind.mail_off_campus or \
+                   annotation_type == "Check Box3" and not data.receive_kind.mail_off_campus:
+                    annotation_obj.update({
+                        NameObject("/V"): NameObject("/Yes"),
+                        NameObject("/AS"): NameObject("/Yes")
+                    })
+        else:
+            raise KeyError("pdf has no checkboxes; something is wrong!")
+
+        if not found_check_box_1 or not found_check_box_3:
+            raise KeyError("could not find checkboxes; pdf may be malformed")
+    else:
+        raise AssertionError("Invalid value {}; exhaustive case".format(data.receive_kind))
+
+    # ------------------------
+    # validation
+
+    fields = reader.get_form_text_fields()
+    for key in relevant_fields.keys():
+        if key not in fields:
+            raise KeyError("pdf form did not have necessary key")
+
+    # ------------------------
+    # final writing
+
+    if not os.path.exists(tmp_dir):
+        os.makedirs(tmp_dir)
+
+    c = canvas.Canvas(tmp_dir + "watermark.pdf")
+    c.drawString(335, 702, "Computing Science Student Society")
+    if watermark_active:
+        c.drawImage(small_text_logo, 429, 725, 848 / 6, 174 / 6)
+    c.save()
+
+    # NOTE: what happens when multiple requests come in at once?
+    # I guess the file acts as a mutex & blocks one of the requests. This degraded performance is
+    # okay! We are using python anyways.
+    csss_text_logo = PdfReader(open(tmp_dir + "watermark.pdf", "rb"))
+    page.merge_page(csss_text_logo.pages[0])
+
+    writer = PdfWriter()
+    writer.add_page(page)
+    writer.update_page_form_field_values(writer.pages[0], relevant_fields)
+
+    # byte stream
+    stream = io.BytesIO()
+    writer.write(stream)
+
+    print("len: {}".format(len(stream.getvalue())))
+    return stream.getvalue()
+
+
+# generate cheque req from grant
+def generate_grant_cheque_req():
+    pass
+
+
+# generate invoice
+def generate_invoice():
+    pass
+
+
+# generate travel & conference funding request
+def generate_travel_conference_funding_req():
+    pass
+
+
+# search through motions from the past n years for keywords
+def search_motions(keyword, threshold=0, nyears=2):
+    # 1. fetch all the pdfs (report how that's going...)
+    # pypdf to read stuff
+
+    # 2. parse them for text
+    # TODO: check if pip install thefuzz[speedup] actually helps & by how much?
+    # thefuzz to match text with keyword
+
+    # 3. return all matches over a certain similarity threshold
+    pass
+
+
+# ------------------------
+# testing
+
+if __name__ == "__main__":

```

Comment:
I assume this will be deleted before this PR is published?

Reply:
I actually planned to leave this for testing purposes. We don't have a formal test suite & I'm pretty fine with leaving it that way

Reply:
ah, no. I applaud you for adding tests and its something I defintely plan to do but add some django style tests

Reply:
ah, sure thing! I'll remove this just before I mark the pr as ready